File: compilers/zig/src/tests/conformance_goldens.zig
//! Byte-identical golden diff harness for the Zig compiler.
//!
//! For every directory under `conformance/tests/`, this test:
//!   1. Locates the Zig-format source file (`*.runar.zig`)
//!   2. Invokes the Zig compiler binary (`zig-out/bin/runar-zig`) via subprocess
//!      with `--emit-ir --disable-constant-folding` and `--hex --disable-constant-folding`
//!   3. Canonicalizes the ANF IR JSON (sort keys, strip `sourceLoc`, 2-space indent)
//!   4. Asserts byte-for-byte equality against `expected-ir.json` and `expected-script.hex`
//!
//! The canonicalization strategy mirrors
//! `conformance/runner/runner.ts::canonicalizeJson`.
//!
//! The binary is resolved via the `RUNAR_ZIG_BIN` environment variable, and
//! falls back to `zig-out/bin/runar-zig` relative to CWD if unset. Run via:
//!   zig build          # builds zig-out/bin/runar-zig
//!   zig build conformance-goldens   # runs this test suite

const std = @import("std");

const conformance_base = "../../conformance/tests/";

/// Resolve a path to the built runar-zig binary.
/// Preference: RUNAR_ZIG_BIN env var, then zig-out/bin/runar-zig, then PATH.
fn resolveZigBinary(allocator: std.mem.Allocator) ![]const u8 {
    if (std.testing.environ.getPosix("RUNAR_ZIG_BIN")) |bin| {
        return try allocator.dupe(u8, bin);
    }
    return try allocator.dupe(u8, "zig-out/bin/runar-zig");
}

/// Resolve the Zig-format source file for a conformance fixture.
///
/// Mirrors the TS runner (`conformance/runner/runner.ts`):
///   1. If `source.json` exists and has a `.runar.zig` entry in `sources`,
///      resolve it relative to the fixture directory. Use it if it exists.
///   2. Otherwise fall back to the first `*.runar.zig` file in the fixture dir.
///
/// Returned slice is allocated with `allocator`.
fn findZigSource(allocator: std.mem.Allocator, io: std.Io, test_dir: []const u8) !?[]u8 {
    // (1) source.json lookup
    const config_path = try std.fmt.allocPrint(allocator, "{s}/source.json", .{test_dir});
    defer allocator.free(config_path);
    if (std.Io.Dir.cwd().readFileAlloc(io, config_path, allocator, .limited(1 * 1024 * 1024))) |raw| {
        defer allocator.free(raw);
        if (std.json.parseFromSlice(std.json.Value, allocator, raw, .{})) |parsed| {
            defer parsed.deinit();
            switch (parsed.value) {
                .object => |root| {
                    if (root.get("sources")) |sources_val| {
                        switch (sources_val) {
                            .object => |sources| {
                                if (sources.get(".runar.zig")) |v| {
                                    switch (v) {
                                        .string => |rel| {
                                            const resolved = try std.fmt.allocPrint(allocator, "{s}/{s}", .{ test_dir, rel });
                                            // Confirm existence via stat; fall through on failure.
                                            if (std.Io.Dir.cwd().statFile(io, resolved, .{})) |_| {
                                                return resolved;
                                            } else |_| {
                                                allocator.free(resolved);
                                            }
                                        },
                                        else => {},
                                    }
                                }
                            },
                            else => {},
                        }
                    }
                },
                else => {},
            }
        } else |_| {}
    } else |_| {}

    // (2) glob fallback in fixture directory
    var dir = std.Io.Dir.cwd().openDir(io, test_dir, .{ .iterate = true }) catch return null;
    defer dir.close(io);

    var iter = dir.iterate();
    var names: std.ArrayListUnmanaged([]u8) = .empty;
    defer {
        for (names.items) |n| allocator.free(n);
        names.deinit(allocator);
    }

    while (try iter.next(io)) |entry| {
        if (entry.kind != .file) continue;
        if (std.mem.endsWith(u8, entry.name, ".runar.zig")) {
            try names.append(allocator, try allocator.dupe(u8, entry.name));
        }
    }
    if (names.items.len == 0) return null;

    // Sort for deterministic selection
    std.mem.sort([]u8, names.items, {}, struct {
        fn lessThan(_: void, a: []u8, b: []u8) bool {
            return std.mem.lessThan(u8, a, b);
        }
    }.lessThan);

    return try std.fmt.allocPrint(allocator, "{s}/{s}", .{ test_dir, names.items[0] });
}

/// Run the Zig compiler binary with the given flags.
/// Returns stdout on success (caller frees); returns error + prints stderr on failure.
fn runCompiler(
    allocator: std.mem.Allocator,
    io: std.Io,
    binary: []const u8,
    source_path: []const u8,
    flags: []const []const u8,
) ![]u8 {
    var argv: std.ArrayListUnmanaged([]const u8) = .empty;
    defer argv.deinit(allocator);
    try argv.append(allocator, binary);
    try argv.append(allocator, "--source");
    try argv.append(allocator, source_path);
    for (flags) |f| try argv.append(allocator, f);

    const result = try std.process.run(allocator, io, .{
        .argv = argv.items,
        .stdout_limit = .limited(10 * 1024 * 1024),
    });
    defer allocator.free(result.stderr);

    switch (result.term) {
        .exited => |code| if (code != 0) {
            defer allocator.free(result.stdout);
            std.debug.print("  runar-zig exit code {d}: {s}\n", .{ code, result.stderr });
            return error.CompilerExitNonZero;
        },
        else => {
            defer allocator.free(result.stdout);
            std.debug.print("  runar-zig abnormal termination\n", .{});
            return error.CompilerAbnormalExit;
        },
    }
    return result.stdout;
}

/// Recursively canonicalize a std.json.Value: strip `sourceLoc`, sort keys,
/// re-serialize with 2-space indent.
fn canonicalizeJson(allocator: std.mem.Allocator, json_str: []const u8) ![]u8 {
    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json_str, .{});
    defer parsed.deinit();

    var buf: std.ArrayListUnmanaged(u8) = .empty;
    errdefer buf.deinit(allocator);
    try writeValue(allocator, &buf, parsed.value, 0);
    return buf.toOwnedSlice(allocator);
}

fn writeIndent(allocator: std.mem.Allocator, buf: *std.ArrayListUnmanaged(u8), depth: usize) !void {
    var i: usize = 0;
    while (i < depth) : (i += 1) {
        try buf.appendSlice(allocator, "  ");
    }
}

fn writeValue(
    allocator: std.mem.Allocator,
    buf: *std.ArrayListUnmanaged(u8),
    v: std.json.Value,
    depth: usize,
) !void {
    switch (v) {
        .null => try buf.appendSlice(allocator, "null"),
        .bool => |b| try buf.appendSlice(allocator, if (b) "true" else "false"),
        .integer => |i| {
            var tmp: [32]u8 = undefined;
            const s = try std.fmt.bufPrint(&tmp, "{d}", .{i});
            try buf.appendSlice(allocator, s);
        },
        .float => |f| {
            var tmp: [64]u8 = undefined;
            const s = try std.fmt.bufPrint(&tmp, "{d}", .{f});
            try buf.appendSlice(allocator, s);
        },
        .number_string => |s| try buf.appendSlice(allocator, s),
        .string => |s| {
            try buf.append(allocator, '"');
            try writeJsonEscapedString(allocator, buf, s);
            try buf.append(allocator, '"');
        },
        .array => |arr| {
            if (arr.items.len == 0) {
                try buf.appendSlice(allocator, "[]");
                return;
            }
            try buf.append(allocator, '[');
            for (arr.items, 0..) |item, i| {
                try buf.append(allocator, '\n');
                try writeIndent(allocator, buf, depth + 1);
                try writeValue(allocator, buf, item, depth + 1);
                if (i + 1 < arr.items.len) try buf.append(allocator, ',');
            }
            try buf.append(allocator, '\n');
            try writeIndent(allocator, buf, depth);
            try buf.append(allocator, ']');
        },
        .object => |obj| {
            // Collect non-sourceLoc keys, sort them
            var keys: std.ArrayListUnmanaged([]const u8) = .empty;
            defer keys.deinit(allocator);
            var it = obj.iterator();
            while (it.next()) |kv| {
                if (std.mem.eql(u8, kv.key_ptr.*, "sourceLoc")) continue;
                try keys.append(allocator, kv.key_ptr.*);
            }
            std.mem.sort([]const u8, keys.items, {}, struct {
                fn lessThan(_: void, a: []const u8, b: []const u8) bool {
                    return std.mem.lessThan(u8, a, b);
                }
            }.lessThan);

            if (keys.items.len == 0) {
                try buf.appendSlice(allocator, "{}");
                return;
            }
            try buf.append(allocator, '{');
            for (keys.items, 0..) |k, i| {
                try buf.append(allocator, '\n');
                try writeIndent(allocator, buf, depth + 1);
                try buf.append(allocator, '"');
                try writeJsonEscapedString(allocator, buf, k);
                try buf.appendSlice(allocator, "\": ");
                try writeValue(allocator, buf, obj.get(k).?, depth + 1);
                if (i + 1 < keys.items.len) try buf.append(allocator, ',');
            }
            try buf.append(allocator, '\n');
            try writeIndent(allocator, buf, depth);
            try buf.append(allocator, '}');
        },
    }
}

fn writeJsonEscapedString(
    allocator: std.mem.Allocator,
    buf: *std.ArrayListUnmanaged(u8),
    s: []const u8,
) !void {
    for (s) |c| {
        switch (c) {
            '"' => try buf.appendSlice(allocator, "\\\""),
            '\\' => try buf.appendSlice(allocator, "\\\\"),
            0x08 => try buf.appendSlice(allocator, "\\b"),
            0x0c => try buf.appendSlice(allocator, "\\f"),
            '\n' => try buf.appendSlice(allocator, "\\n"),
            '\r' => try buf.appendSlice(allocator, "\\r"),
            '\t' => try buf.appendSlice(allocator, "\\t"),
            else => {
                if (c < 0x20) {
                    var tmp: [8]u8 = undefined;
                    const esc = try std.fmt.bufPrint(&tmp, "\\u{x:0>4}", .{c});
                    try buf.appendSlice(allocator, esc);
                } else {
                    try buf.append(allocator, c);
                }
            },
        }
    }
}

/// Strip ASCII whitespace from `input`, lowercase hex digits, return new slice.
fn normalizeHex(allocator: std.mem.Allocator, input: []const u8) ![]u8 {
    var out: std.ArrayListUnmanaged(u8) = .empty;
    errdefer out.deinit(allocator);
    try out.ensureTotalCapacity(allocator, input.len);
    for (input) |c| {
        if (c == ' ' or c == '\t' or c == '\n' or c == '\r') continue;
        const lc = if (c >= 'A' and c <= 'Z') c + 32 else c;
        try out.append(allocator, lc);
    }
    return out.toOwnedSlice(allocator);
}

// ============================================================================
// Shared state for the aggregate summary test
// ============================================================================

const FailureKind = enum { compile_ir, compile_hex, canon_ir, ir_mismatch, script_mismatch };

const Failure = struct {
    name: []const u8,
    kind: FailureKind,
    detail: []const u8,
    expected: ?[]const u8 = null,
    actual: ?[]const u8 = null,
};

fn printShortDiff(expected: []const u8, actual: []const u8) void {
    var exp_lines_iter = std.mem.splitScalar(u8, expected, '\n');
    var act_lines_iter = std.mem.splitScalar(u8, actual, '\n');
    var shown: usize = 0;
    var line: usize = 1;
    while (true) : (line += 1) {
        const e = exp_lines_iter.next();
        const a = act_lines_iter.next();
        if (e == null and a == null) break;
        const el = e orelse "<EOF>";
        const al = a orelse "<EOF>";
        if (!std.mem.eql(u8, el, al)) {
            std.debug.print("    line {d}:\n      - expected: {s}\n      + actual:   {s}\n", .{ line, el, al });
            shown += 1;
            if (shown >= 12) {
                std.debug.print("    ... (truncated)\n", .{});
                return;
            }
        }
    }
}

fn printScriptDiff(expected: []const u8, actual: []const u8) void {
    const min_len = @min(expected.len, actual.len);
    var first_diff: usize = min_len;
    var i: usize = 0;
    while (i < min_len) : (i += 1) {
        if (expected[i] != actual[i]) {
            first_diff = i;
            break;
        }
    }
    const lo = if (first_diff >= 20) first_diff - 20 else 0;
    const exp_hi = @min(first_diff + 20, expected.len);
    const act_hi = @min(first_diff + 20, actual.len);
    std.debug.print(
        "  expected {d} hex chars, actual {d} hex chars\n  first diff at hex offset {d} (byte {d})\n  expected: ...{s}...\n  actual:   ...{s}...\n",
        .{ expected.len, actual.len, first_diff, first_diff / 2, expected[lo..exp_hi], actual[lo..act_hi] },
    );
}

// ============================================================================
// The single conformance test — iterates all 40 fixture directories
// ============================================================================

test "conformance-goldens: zig compiler produces byte-identical output for all fixtures" {
    const allocator = std.testing.allocator;
    const io = std.testing.io;

    const binary = try resolveZigBinary(allocator);
    defer allocator.free(binary);

    var dir = std.Io.Dir.cwd().openDir(io, conformance_base, .{ .iterate = true }) catch |err| {
        std.debug.print("cannot open {s}: {s}\n", .{ conformance_base, @errorName(err) });
        return err;
    };
    defer dir.close(io);

    // Gather fixture dirs
    var fixture_names: std.ArrayListUnmanaged([]u8) = .empty;
    defer {
        for (fixture_names.items) |n| allocator.free(n);
        fixture_names.deinit(allocator);
    }
    var it = dir.iterate();
    while (try it.next(io)) |entry| {
        if (entry.kind != .directory) continue;
        try fixture_names.append(allocator, try allocator.dupe(u8, entry.name));
    }
    std.mem.sort([]u8, fixture_names.items, {}, struct {
        fn lessThan(_: void, a: []u8, b: []u8) bool {
            return std.mem.lessThan(u8, a, b);
        }
    }.lessThan);

    var passed: std.ArrayListUnmanaged([]const u8) = .empty;
    defer passed.deinit(allocator);
    var missing: std.ArrayListUnmanaged([]const u8) = .empty;
    defer missing.deinit(allocator);
    var failures: std.ArrayListUnmanaged(Failure) = .empty;
    defer {
        for (failures.items) |f| {
            allocator.free(f.detail);
            if (f.expected) |e| allocator.free(e);
            if (f.actual) |a| allocator.free(a);
        }
        failures.deinit(allocator);
    }

    for (fixture_names.items) |name| {
        const test_dir = try std.fmt.allocPrint(allocator, "{s}{s}", .{ conformance_base, name });
        defer allocator.free(test_dir);

        const source_opt = try findZigSource(allocator, io, test_dir);
        if (source_opt == null) {
            try missing.append(allocator, name);
            continue;
        }
        const source_path = source_opt.?;
        defer allocator.free(source_path);

        // Step 1: emit IR
        const ir_flags = [_][]const u8{ "--emit-ir", "--disable-constant-folding" };
        const ir_raw = runCompiler(allocator, io, binary, source_path, &ir_flags) catch |err| {
            const detail = try std.fmt.allocPrint(allocator, "emit-ir failed: {s}", .{@errorName(err)});
            try failures.append(allocator, .{ .name = name, .kind = .compile_ir, .detail = detail });
            continue;
        };
        defer allocator.free(ir_raw);

        // Step 2: emit hex
        const hex_flags = [_][]const u8{ "--hex", "--disable-constant-folding" };
        const hex_raw = runCompiler(allocator, io, binary, source_path, &hex_flags) catch |err| {
            const detail = try std.fmt.allocPrint(allocator, "emit-hex failed: {s}", .{@errorName(err)});
            try failures.append(allocator, .{ .name = name, .kind = .compile_hex, .detail = detail });
            continue;
        };
        defer allocator.free(hex_raw);

        // Canonicalize actual IR
        const actual_ir = canonicalizeJson(allocator, ir_raw) catch |err| {
            const detail = try std.fmt.allocPrint(allocator, "canonicalize actual IR: {s}", .{@errorName(err)});
            try failures.append(allocator, .{ .name = name, .kind = .canon_ir, .detail = detail });
            continue;
        };

        // Compare IR
        const ir_path = try std.fmt.allocPrint(allocator, "{s}/expected-ir.json", .{test_dir});
        defer allocator.free(ir_path);
        const expected_ir_raw = std.Io.Dir.cwd().readFileAlloc(io, ir_path, allocator, .limited(16 * 1024 * 1024)) catch null;
        if (expected_ir_raw) |raw| {
            defer allocator.free(raw);
            const expected_ir = canonicalizeJson(allocator, raw) catch |err| {
                allocator.free(actual_ir);
                const detail = try std.fmt.allocPrint(allocator, "canonicalize expected IR: {s}", .{@errorName(err)});
                try failures.append(allocator, .{ .name = name, .kind = .canon_ir, .detail = detail });
                continue;
            };
            if (!std.mem.eql(u8, actual_ir, expected_ir)) {
                const detail = try std.fmt.allocPrint(allocator, "IR differs (exp {d} / act {d} chars)", .{ expected_ir.len, actual_ir.len });
                try failures.append(allocator, .{
                    .name = name,
                    .kind = .ir_mismatch,
                    .detail = detail,
                    .expected = expected_ir,
                    .actual = actual_ir,
                });
                continue;
            }
            allocator.free(expected_ir);
        }
        allocator.free(actual_ir);

        // Compare hex
        const actual_hex = try normalizeHex(allocator, hex_raw);

        const hex_path = try std.fmt.allocPrint(allocator, "{s}/expected-script.hex", .{test_dir});
        defer allocator.free(hex_path);
        const expected_hex_raw = std.Io.Dir.cwd().readFileAlloc(io, hex_path, allocator, .limited(4 * 1024 * 1024)) catch null;
        if (expected_hex_raw) |raw| {
            defer allocator.free(raw);
            const expected_hex = try normalizeHex(allocator, raw);
            if (!std.mem.eql(u8, actual_hex, expected_hex)) {
                const detail = try std.fmt.allocPrint(allocator, "script differs (exp {d} / act {d})", .{ expected_hex.len, actual_hex.len });
                try failures.append(allocator, .{
                    .name = name,
                    .kind = .script_mismatch,
                    .detail = detail,
                    .expected = expected_hex,
                    .actual = actual_hex,
                });
                continue;
            }
            allocator.free(expected_hex);
        }
        allocator.free(actual_hex);

        try passed.append(allocator, name);
    }

    const total = fixture_names.items.len;
    std.debug.print(
        "\n=== Zig conformance-goldens summary: {d} pass / {d} fail / {d} missing-source (of {d} fixtures) ===\n",
        .{ passed.items.len, failures.items.len, missing.items.len, total },
    );
    if (missing.items.len > 0) {
        std.debug.print("Missing .runar.zig source files:\n", .{});
        for (missing.items) |n| std.debug.print("  - {s}\n", .{n});
    }
    const shown = @min(failures.items.len, 5);
    var i: usize = 0;
    while (i < shown) : (i += 1) {
        const f = failures.items[i];
        std.debug.print("\n--- FAIL: {s} ({s}) ---\n", .{ f.name, @tagName(f.kind) });
        switch (f.kind) {
            .ir_mismatch => {
                std.debug.print("  {s}\n", .{f.detail});
                if (f.expected != null and f.actual != null) {
                    printShortDiff(f.expected.?, f.actual.?);
                }
            },
            .script_mismatch => {
                if (f.expected != null and f.actual != null) {
                    printScriptDiff(f.expected.?, f.actual.?);
                }
            },
            else => {
                std.debug.print("  {s}\n", .{f.detail});
            },
        }
    }
    if (failures.items.len > shown) {
        std.debug.print("\n... and {d} more failures:\n", .{failures.items.len - shown});
        var j: usize = shown;
        while (j < failures.items.len) : (j += 1) {
            std.debug.print("  - {s}\n", .{failures.items[j].name});
        }
    }

    if (failures.items.len > 0) {
        std.debug.print("\n{d} of {d} fixtures failed conformance-goldens\n", .{ failures.items.len, total });
        return error.ConformanceGoldensFailed;
    }
}

File: compilers/rust/tests/conformance_goldens.rs
//! Byte-identical golden diff harness for the Rust compiler.
//!
//! For every directory under `conformance/tests/`, this test:
//!   1. Locates the Rust-format source file (`*.runar.rs`)
//!   2. Compiles it via the Rust compiler (parse -> validate -> typecheck -> ANF -> stack -> emit)
//!   3. Canonicalizes the ANF IR JSON (sort keys, strip `sourceLoc`, 2-space indent)
//!   4. Asserts byte-for-byte equality against `expected-ir.json` and `expected-script.hex`
//!
//! The canonicalization strategy mirrors `conformance/runner/runner.ts::canonicalizeJson`.
//!
//! All failures are collected and reported at the end. A summary prints pass/fail counts
//! and shows concrete diffs for the first 5 failing fixtures (IR / script).

use runar_compiler_rust::{
    compile_from_source_str_with_options, compile_source_str_to_ir_with_options, CompileOptions,
};
use serde_json::Value;
use std::fs;
use std::path::{Path, PathBuf};

fn conformance_tests_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join("conformance")
        .join("tests")
}

/// Resolve the Rust-format source file for a conformance fixture.
///
/// Mirrors the TS runner's logic in `conformance/runner/runner.ts`:
///   1. If `source.json` exists and has a `.runar.rs` entry in `sources`,
///      resolve it relative to the fixture directory. Use it if the path exists.
///   2. Otherwise fall back to the first `*.runar.rs` file in the fixture dir.
fn find_rust_source(test_dir: &Path) -> Option<PathBuf> {
    // (1) source.json lookup
    let config_path = test_dir.join("source.json");
    if config_path.exists() {
        if let Ok(raw) = fs::read_to_string(&config_path) {
            if let Ok(cfg) = serde_json::from_str::<Value>(&raw) {
                if let Some(rel) = cfg
                    .get("sources")
                    .and_then(|s| s.get(".runar.rs"))
                    .and_then(|v| v.as_str())
                {
                    let resolved = test_dir.join(rel);
                    // Canonicalize so the path is absolute and normalized;
                    // existence is implied by canonicalize() succeeding.
                    if let Ok(abs) = fs::canonicalize(&resolved) {
                        return Some(abs);
                    }
                }
            }
        }
    }

    // (2) glob fallback in fixture directory
    let entries = fs::read_dir(test_dir).ok()?;
    for entry in entries.flatten() {
        let path = entry.path();
        if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
            if name.ends_with(".runar.rs") {
                return Some(path);
            }
        }
    }
    None
}

/// Recursively sort object keys and strip `sourceLoc`.
fn canonicalize_value(v: &Value) -> Value {
    match v {
        Value::Object(map) => {
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let mut out = serde_json::Map::new();
            for k in keys {
                if k == "sourceLoc" {
                    continue;
                }
                out.insert(k.clone(), canonicalize_value(&map[k]));
            }
            Value::Object(out)
        }
        Value::Array(arr) => Value::Array(arr.iter().map(canonicalize_value).collect()),
        _ => v.clone(),
    }
}

/// Parse + canonicalize + serialize JSON with 2-space indent.
fn canonicalize_json_str(s: &str) -> Result<String, String> {
    let parsed: Value = serde_json::from_str(s).map_err(|e| format!("JSON parse error: {}", e))?;
    let canonical = canonicalize_value(&parsed);
    serde_json::to_string_pretty(&canonical).map_err(|e| format!("JSON serialize error: {}", e))
}

#[derive(Debug)]
enum FixtureOutcome {
    Pass,
    MissingSource,
    CompileError(String),
    IrMismatch { expected: String, actual: String },
    ScriptMismatch { expected: String, actual: String },
}

fn run_one_fixture(test_dir: &Path) -> FixtureOutcome {
    let source_path = match find_rust_source(test_dir) {
        Some(p) => p,
        None => return FixtureOutcome::MissingSource,
    };

    let source = match fs::read_to_string(&source_path) {
        Ok(s) => s,
        Err(e) => return FixtureOutcome::CompileError(format!("read source: {}", e)),
    };
    let file_name = source_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("contract.rs");

    let opts = CompileOptions {
        disable_constant_folding: true,
        ..Default::default()
    };

    // Step 1: compile to IR
    let anf = match compile_source_str_to_ir_with_options(&source, Some(file_name), &opts) {
        Ok(a) => a,
        Err(e) => return FixtureOutcome::CompileError(format!("source->IR: {}", e)),
    };
    let actual_ir_json = match serde_json::to_string(&anf) {
        Ok(s) => s,
        Err(e) => return FixtureOutcome::CompileError(format!("serialize IR: {}", e)),
    };
    let actual_ir_canonical = match canonicalize_json_str(&actual_ir_json) {
        Ok(s) => s,
        Err(e) => return FixtureOutcome::CompileError(format!("canonicalize actual IR: {}", e)),
    };

    // Step 2: compile to script hex
    let artifact = match compile_from_source_str_with_options(&source, Some(file_name), &opts) {
        Ok(a) => a,
        Err(e) => return FixtureOutcome::CompileError(format!("source->script: {}", e)),
    };
    let actual_script = artifact.script.to_lowercase();
    let actual_script: String = actual_script.chars().filter(|c| !c.is_whitespace()).collect();

    // Step 3: compare against goldens
    let expected_ir_path = test_dir.join("expected-ir.json");
    if expected_ir_path.exists() {
        let raw = match fs::read_to_string(&expected_ir_path) {
            Ok(s) => s,
            Err(e) => return FixtureOutcome::CompileError(format!("read golden IR: {}", e)),
        };
        let expected_canonical = match canonicalize_json_str(&raw) {
            Ok(s) => s,
            Err(e) => return FixtureOutcome::CompileError(format!("canonicalize golden IR: {}", e)),
        };
        if actual_ir_canonical != expected_canonical {
            return FixtureOutcome::IrMismatch {
                expected: expected_canonical,
                actual: actual_ir_canonical,
            };
        }
    }

    let expected_script_path = test_dir.join("expected-script.hex");
    if expected_script_path.exists() {
        let raw = match fs::read_to_string(&expected_script_path) {
            Ok(s) => s,
            Err(e) => return FixtureOutcome::CompileError(format!("read golden script: {}", e)),
        };
        let expected_script: String = raw.to_lowercase().chars().filter(|c| !c.is_whitespace()).collect();
        if actual_script != expected_script {
            return FixtureOutcome::ScriptMismatch {
                expected: expected_script,
                actual: actual_script,
            };
        }
    }

    FixtureOutcome::Pass
}

/// Produce a compact unified-diff summary of two strings (first 40 differing lines).
fn short_diff(expected: &str, actual: &str) -> String {
    let exp_lines: Vec<&str> = expected.lines().collect();
    let act_lines: Vec<&str> = actual.lines().collect();
    let mut out = String::new();
    let max = exp_lines.len().max(act_lines.len());
    let mut shown = 0usize;
    for i in 0..max {
        let e = exp_lines.get(i).copied().unwrap_or("<EOF>");
        let a = act_lines.get(i).copied().unwrap_or("<EOF>");
        if e != a {
            out.push_str(&format!("    line {}:\n      - expected: {}\n      + actual:   {}\n", i + 1, e, a));
            shown += 1;
            if shown >= 12 {
                out.push_str("    ... (truncated)\n");
                break;
            }
        }
    }
    if out.is_empty() {
        out.push_str("    (strings differ but no line diff; likely trailing whitespace)\n");
    }
    out
}

#[test]
fn test_conformance_goldens_rust() {
    let tests_dir = conformance_tests_dir();
    assert!(
        tests_dir.is_dir(),
        "conformance tests directory not found: {}",
        tests_dir.display()
    );

    let mut dirs: Vec<PathBuf> = fs::read_dir(&tests_dir)
        .expect("readdir conformance/tests")
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
        .map(|e| e.path())
        .collect();
    dirs.sort();

    let mut pass: Vec<String> = Vec::new();
    let mut missing: Vec<String> = Vec::new();
    let mut failures: Vec<(String, FixtureOutcome)> = Vec::new();

    for dir in &dirs {
        let name = dir.file_name().unwrap().to_string_lossy().to_string();
        let outcome = run_one_fixture(dir);
        match &outcome {
            FixtureOutcome::Pass => pass.push(name),
            FixtureOutcome::MissingSource => missing.push(name),
            _ => failures.push((name, outcome)),
        }
    }

    // Print summary
    let total = dirs.len();
    let pass_count = pass.len();
    let miss_count = missing.len();
    let fail_count = failures.len();
    println!(
        "\n=== Rust conformance-goldens summary: {} pass / {} fail / {} missing-source (of {} fixtures) ===",
        pass_count, fail_count, miss_count, total
    );
    if !missing.is_empty() {
        println!("Missing .runar.rs source files:");
        for n in &missing {
            println!("  - {}", n);
        }
    }
    // Show up to the first 5 failures with concrete diffs
    let shown_failures: Vec<&(String, FixtureOutcome)> = failures.iter().take(5).collect();
    for (name, outcome) in &shown_failures {
        println!("\n--- FAIL: {} ---", name);
        match outcome {
            FixtureOutcome::CompileError(msg) => {
                println!("  compile error: {}", msg);
            }
            FixtureOutcome::IrMismatch { expected, actual } => {
                println!("  IR mismatch (expected {} chars, actual {} chars):", expected.len(), actual.len());
                print!("{}", short_diff(expected, actual));
            }
            FixtureOutcome::ScriptMismatch { expected, actual } => {
                println!(
                    "  script mismatch: expected {} hex chars, actual {} hex chars",
                    expected.len(),
                    actual.len()
                );
                let min_len = expected.len().min(actual.len());
                let mut first_diff = min_len;
                for i in 0..min_len {
                    if expected.as_bytes()[i] != actual.as_bytes()[i] {
                        first_diff = i;
                        break;
                    }
                }
                let lo = first_diff.saturating_sub(20);
                let exp_hi = (first_diff + 20).min(expected.len());
                let act_hi = (first_diff + 20).min(actual.len());
                println!(
                    "  first diff at hex offset {} (byte {})",
                    first_diff,
                    first_diff / 2
                );
                println!("  expected: ...{}...", &expected[lo..exp_hi]);
                println!("  actual:   ...{}...", &actual[lo..act_hi]);
            }
            _ => {}
        }
    }
    if failures.len() > shown_failures.len() {
        println!("\n... and {} more failures:", failures.len() - shown_failures.len());
        for (name, _) in &failures[shown_failures.len()..] {
            println!("  - {}", name);
        }
    }

    assert!(
        failures.is_empty(),
        "{} of {} fixtures failed (see stdout for details)",
        failures.len(),
        total
    );
}
